From 18f33ac86aaec31be0eda5335df78c1fb8b45459 Mon Sep 17 00:00:00 2001
From: S-trace <S-trace@list.ru>
Date: Wed, 28 Feb 2018 02:21:59 +0300
Subject: [PATCH 9/9] [KEYSTORE] KeyStore.java: Catch NullPointerExceptions

Fixes sdcard boot failure:
02-28 02:00:18.370 13545 13545 W SystemServer: ***********************************************
02-28 02:00:18.371 13545 13545 E SystemServer: BOOT FAILURE making Connectivity Service ready
02-28 02:00:18.371 13545 13545 E SystemServer: java.lang.NullPointerException: Attempt to invoke interface method 'int android.security.IKeystoreService.exist(java.lang.String, int)' on a null object reference
02-28 02:00:18.371 13545 13545 E SystemServer:  at android.security.KeyStore.contains(KeyStore.java:200)
02-28 02:00:18.371 13545 13545 E SystemServer:  at android.security.KeyStore.contains(KeyStore.java:208)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.net.LockdownVpnTracker.isEnabled(LockdownVpnTracker.java:90)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.ConnectivityService.updateLockdownVpn(ConnectivityService.java:3712)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.ConnectivityService.systemReady(ConnectivityService.java:1759)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.SystemServer$2.run(SystemServer.java:1526)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.am.ActivityManagerService.systemReady(ActivityManagerService.java:13578)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.SystemServer.startOtherServices(SystemServer.java:1466)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.SystemServer.run(SystemServer.java:368)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.server.SystemServer.main(SystemServer.java:237)
02-28 02:00:18.371 13545 13545 E SystemServer:  at java.lang.reflect.Method.invoke(Native Method)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
02-28 02:00:18.371 13545 13545 E SystemServer:  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)

Change-Id: I668cc4e63de72a230344239703cc09d634dc8682
---
 keystore/java/android/security/KeyStore.java | 105 +++++++++++++++++++++++++++
 1 file changed, 105 insertions(+)

diff --git a/keystore/java/android/security/KeyStore.java b/keystore/java/android/security/KeyStore.java
index 00d786a9362..9f6240080ca 100644
--- a/keystore/java/android/security/KeyStore.java
+++ b/keystore/java/android/security/KeyStore.java
@@ -161,6 +161,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
 
@@ -178,6 +181,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -188,6 +194,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -201,6 +210,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -217,6 +229,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
 
@@ -230,6 +245,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -245,6 +263,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -270,6 +291,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -286,6 +310,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -301,6 +328,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -310,6 +340,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -319,6 +352,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
 
@@ -328,6 +364,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -337,6 +376,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -346,6 +388,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -364,6 +409,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return -1L;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return -1L;
         }
     }
 
@@ -377,6 +425,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -391,6 +442,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -400,6 +454,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -413,6 +470,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -423,6 +483,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -438,6 +501,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -454,6 +520,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -469,6 +538,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
     public ExportResult exportKey(String alias, int format, KeymasterBlob clientId,
@@ -483,6 +555,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
 
@@ -497,6 +572,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
 
@@ -507,6 +585,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return null;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return null;
         }
     }
 
@@ -520,6 +601,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -535,6 +619,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -551,6 +638,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -571,6 +661,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return false;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return false;
         }
     }
 
@@ -587,6 +680,9 @@ public class KeyStore {
             mBinder.onUserAdded(userId, parentId);
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return;
         }
     }
 
@@ -609,6 +705,9 @@ public class KeyStore {
             mBinder.onUserRemoved(userId);
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return;
         }
     }
 
@@ -623,6 +722,9 @@ public class KeyStore {
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
             return SYSTEM_ERROR;
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return SYSTEM_ERROR;
         }
     }
 
@@ -634,6 +736,9 @@ public class KeyStore {
             mBinder.onDeviceOffBody();
         } catch (RemoteException e) {
             Log.w(TAG, "Cannot connect to keystore", e);
+        } catch (NullPointerException e) {
+            Log.w(TAG, "mBinder is null", e);
+            return;
         }
     }
 
-- 
2.14.1

